community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -470,6 +472,92 @@ def update_topology(lkp: util.Lookup) -> None:
         log.debug("Topology configuration updated. Reconfiguring Slurm.")
         util.scontrol_reconfigure(lkp)
 
+
+def delete_reservation(reservation_name: str) -> None:
+    util.run(f"{lkp.scontrol} delete reservation {reservation_name}")
+
+
+def create_reservation(reservation_name: str, node: str, starttime: str) -> None:
+    # Format time to be compatible with slurm reservation.
+    start_time=datetime.strptime(starttime, '%Y-%m-%dT%H:%M:%S%z')
+    formatted_start_time = start_time.strftime('%Y-%m-%dT%H:%M:%S')
+    util.run(f"{lkp.scontrol} create reservation user=slurm starttime={formatted_start_time} duration=180 nodes={node} reservationname={reservation_name}")
+
+
+def get_slurm_reservation_maintenance() -> Dict[str, str]:
+    res = util.run(f"{lkp.scontrol} show reservation --json")
+    all_reservations = json.loads(res.stdout)
+    reservation_map = {}
+
+    for reservation in all_reservations['reservations']:
+        reservation_name = reservation['name']
+        nodes = reservation['node_list']
+        time = reservation['start_time']['number']
+
+        if reservation_name is None or nodes is None or time is None:
+          continue
+
+        # There should be 1 reservation per node per maintenance.
+        if reservation['node_count'] != 1:
+          continue
+
+        # Check if the reservation is for scheduled maintenance.
+        maintenance_reservation = nodes + "_maintenance"
+        if reservation_name != maintenance_reservation:
+          continue
+
+        reservation_map[reservation_name] = time
+
+    return reservation_map
+
+
+def get_upcoming_maintenance(lkp: util.Lookup) -> Dict[str, Tuple[str, datetime]]:
+    upc_maint_map = {}
+
+    for node, properties in lkp.instances().items():
+        if 'upcomingMaintenance' in properties:
+          upc_maint_map[node + "_maintenance"] = (node, properties['upcomingMaintenance']['startTimeWindow']['earliest'])
```

Comment:
AFAIK.  `properties['upcomingMaintenance']['startTimeWindow']['earliest']` is not a `datetime`, but string in RFC3339 format. Is it correct? Please parse the string and update (or keep) the type annotation.

Reply:
Updated both to datetime object until we create reservation.